src/relations/mysql.py

```diff
@@ -0,0 +1,146 @@
+            for key, value in relation_data.items():
+                if relation_databag[self.charm.unit].get(key) != value:
+                    relation_databag[self.charm.unit][key] = value
+
+            # Assign the cluster primary's address as the database host
+            primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+            relation_databag[self.charm.unit]["host"] = primary_address
+
+    def _on_mysql_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handle the legacy 'mysql' relation created event.
+
+        Will set up the database and the scoped application user. The connection
+        data (relation data) is then copied into the peer relation databag (to
+        be copied over to the new leader unit's databag in case of a new leader
+        being elected).
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        # Wait until on-config-changed event is executed
+        # (wait for root password to have been set)
+        if not self.charm._is_peer_data_set:
+            event.defer()
+            return
+
+        logger.warning("DEPRECATION WARNING - `mysql` is a legacy interface")
+
+        username = self.charm.config.get("mysql-interface-user")
+        database = self.charm.config.get("mysql-interface-database")
+
+        if not username or not database:
+            event.defer()
+            return
+
+        # Only execute if the application user does not exist
+        if self.charm._mysql.does_mysql_user_exist(username, "%"):
+            return
+
+        password = self._get_or_set_password_in_peer_databag(username)
+
+        self.charm._mysql.create_application_database_and_scoped_user(
+            database,
+            username,
+            password,
+            "%",
+            "mysql-legacy-relation",
+        )
+
+        primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
+        updates = {
+            "database": database,
+            "host": primary_address,
+            "password": password,
+            "port": "3306",
+            "root_password": self.charm._peers.data[self.charm.app]["root-password"],
+            "user": username,
+        }
+
+        event.relation.data[self.charm.unit].update(updates)
+
+        # Store the relation data into the peer relation databag
+        self.charm._peers.data[self.charm.app]["mysql_relation_data"] = json.dumps(updates)
+
+    def _on_mysql_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle the 'mysql' legacy relation broken event.
+
+        Delete the application user created in the relation created
+        event handler.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.warning("DEPRECATION WARNING - `mysql` is a legacy interface")
+
+        self.charm._mysql.delete_users_for_unit("mysql-legacy-relation")
```

Comment:
same here `MySQLDeleteUsersForUnitError `, should we defer or retry or maybe go into blocked status?